src/palindrom.js

```diff
@@ -193,319 +187,317 @@ var Palindrom = (function() {
+          headers
+        });
+      } else {
+        requestPromise = axios.patch(url, data, {
+          headers
+        });
+      }
+      requestPromise
+        .then(res => {
+          this.handleResponseHeader(res);
+          callback && callback.call(this.palindrom, res, method);
+        })
+        .catch(error => {
+          const res = error.response;
+
+          if (res) {
+            var statusCode = res.status;
+            var statusText = res.statusText;
+            var reason = res.data;
+          } else {
+            // no sufficient error information, we need to create on our own
+            var statusCode = -1;
+            var statusText = `An unknown network error has occurred. Raw message: ${error.message}`;
+            var reason = 'Maybe you lost connection with the server';
+            // log it for verbosity
+            console.error(error);
+          }
+          this.onFatalError(
+            {
+              statusCode,
+              statusText,
+              reason
+            },
+            url,
+            method
+          );
+        });
+
+      this.onSend(data, url, method);
     }
-    requestPromise
-      .then(function(res) {
-        that.handleResponseHeader(res);
-        callback && callback.call(that.palindrom, res, method);
-      })
-      .catch(function(error) {
-        const res = error.response;
-
-        if (res) {
-          var statusCode = res.status;
-          var statusText = res.statusText;
-          var reason = res.data;
-        } else {
-          // no sufficient error information, we need to create on our own
-          var statusCode = -1;
-          var statusText = `An unknown network error has occurred. Raw message: ${error.message}`;
-          var reason = 'Maybe you lost connection with the server';
-          // log it for verbosity
-          console.error(error);
-        }
-        that.onFatalError(
-          {
-            statusCode,
-            statusText,
-            reason
-          },
-          url,
-          method
-        );
-      });
+  }
+  // TODO: auto-configure here #38 (tomalec)
+  function establish(network, url, body, bootstrap) {
+    return network.xhr(url, 'application/json', body, res => {
+      bootstrap(res.data);
+      if (network.useWebSocket) {
+        network.webSocketUpgrade(network.onSocketOpened);
+      }
+    });
+  }
 
-    this.onSend(data, url, method);
-  };
+  function closeWsIfNeeded(network) {
+    if (network._ws) {
+      network._ws.onclose = () => {};
+      network._ws.close();
+      network._ws = null;
+    }
+  }
 
   /**
    * Non-queuing object that conforms JSON-Patch-Queue API
    * @param {Object} obj target object where patches are applied
    * @param {Function} apply function to apply received patch, must return the object in its final state
    */
-  function NoQueue(obj, apply) {
-    this.obj = obj;
-    this.apply = apply;
+  class NoQueue {
```

Comment:
Do you have benchmarks before and after this merge? As I'm not so sure if classes are faster than pure objects + prototypes.

Reply:
Nobody claims that classes are faster than puer objects + prototypes. I think the choice was made for cleaner design.

This article shows classes can be actually faster than objects + prototypes: http://incaseofstairs.com/six-speed/

Reply:
TBH I did see @warpech's link but only after I did this change. However, I did consider performance and my motivation was:

1. Construction happens once per page load, even if it took 200% more time (which is not the case) it can hardly add 1ms to page load. After that they don't affect the experience.
2. Classes are future-friendly and give JS engines much more room for [static analysis](https://en.wikipedia.org/wiki/Static_program_analysis). That's why I love `getter` and `setters` much more than `defineProperty`. You need to execute some JS in order to process `Object.defineProperty`, but you don't need to execute any JS to define `get prop() {}`. And this should make it much faster when JS engines evolve. This is the same reason I don't like to mix classes with `className.prototype.methed = function() {}` because I think this notation misses the point of ES6 classes.